src/sagas/background-refresh.js

```diff
@@ -0,0 +1,230 @@
+
+import AppConfiguration from '_/config'
+import { isNumber } from '_/utils'
+import { delay } from './utils'
+
+import {
+  fetchByPage,
+  fetchPools,
+  fetchSinglePool,
+  fetchSingleVm,
+  fetchVms,
+  selectVmDetail,
+} from './index'
+import { getConsoleOptions } from './console'
+import { fetchIsoFiles } from './storageDomains'
+
+/**
+ * Change the background refresh type based on the page type, and force a refresh.
+ *
+ * This should be done at the time of navigation to the page, typically by the page router.
+ */
+function* changePage (action) {
+  yield put(Actions.stopSchedulerFixedDelay())
+  yield put(Actions.setCurrentPage(action.payload))
+  yield put(Actions.refresh({
+    onNavigation: true,
+    shallowFetch: true,
+  }))
+  yield put(Actions.startSchedulerFixedDelay())
+}
+
+/**
+ * Invoke the correct refresh function based on the app's current page type.
+ */
+function* refreshData (action) {
+  const currentPage = yield select(state => state.config.get('currentPage'))
+  const refreshType =
+    currentPage.type === C.NO_REFRESH_TYPE ? null
+      : currentPage.type === undefined ? C.LIST_PAGE_TYPE
+        : currentPage.type
+
+  console.info('refreshData() 🡒', 'refreshType:', refreshType, 'currentPage:', currentPage, 'payload:', action.payload)
+  if (refreshType) {
+    yield pagesRefreshers[refreshType](Object.assign({ id: currentPage.id }, action.payload))
+  }
+  console.info('refreshData() 🡒 finished')
+}
+
+const pagesRefreshers = {
+  [C.LIST_PAGE_TYPE]: refreshListPage,
+  [C.DETAIL_PAGE_TYPE]: refreshDetailPage,
+  [C.CREATE_PAGE_TYPE]: refreshCreatePage,
+  [C.CONSOLE_PAGE_TYPE]: refreshConsolePage,
+}
+
+function* getIdsByType (type) {
+  const ids = Array.from(yield select(state => state.vms.get(type).keys()))
+  return ids
+}
+
+function* refreshListPage ({ shallowFetch, onNavigation, onSchedule }) {
+  const [ vmsPage, poolsPage ] = yield select(st => [ st.vms.get('vmsPage'), st.vms.get('poolsPage') ])
+
+  // Special case for the very first `refreshListPage` of the App .. use fetchByPage()!
+  if (vmsPage === 0 && poolsPage === 0) {
+    yield call(fetchByPage)
+    return
+  }
+
+  const [ vmsCount, poolsCount ] = yield all([
+    call(function* () {
+      // refresh VMs and remove any that haven't been refreshed
+      if (vmsPage > 0) {
+        const fetchedVmIds = yield fetchVms(Actions.getVmsByCount({
+          count: vmsPage * AppConfiguration.pageLimit,
+          shallowFetch,
+        }))
+
+        const vmsIds = yield getIdsByType('vms')
+        const fetchedDirectlyVmIds =
+          (yield all(
+            vmsIds
+              .filter(vmId => !fetchedVmIds.includes(vmId))
+              .map(vmId => call(fetchSingleVm, Actions.getSingleVm({ vmId, shallowFetch })))
+          ))
+            .reduce((vmIds, vm) => { if (vm) vmIds.push(vm.id); return vmIds }, [])
+
+        yield put(Actions.removeMissingVms({ vmIdsToPreserve: [ ...fetchedVmIds, ...fetchedDirectlyVmIds ] }))
```

Comment:
Should we update paging data here?
The actual number of vm's might be more or less then `vmsPage * AppConfiguration.pageLimit`.

Reply:
Yes, but it happen below the `all()` call.  The next line here `return yield select` is what this sub-saga returns.  Same thing happens in the pool sub-saga.

Lines 137-140 does the paging update to VMs and Pool at the same time after they've both been fully updated.

Reply:
Thanks for clarification!